### PR TITLE
Pass ReferenceCell::Type to QProjection

### PR DESCRIPTION
**Pull request by peterrum**
... and move the implementation of `QProjection` to a new file `qprojection.cc`

### CODE REVIEW

File: include/deal.II/fe/fe_base.h [hunk 1]
```diff
@@ -219,6 +219,13 @@ class FiniteElementData
    */
   static const unsigned int dimension = dim;
 
+private:
+  /**
+   * Reference cell type.
+   */
+  const ReferenceCell::Type cell_type;
+
+public:
```
Comment (peterrum) [inline include/deal.II/fe/fe_base.h:228]: From #10717.

**Comment by peterrum** [comment]
This PR does not change any functionality. It only adds an additional parameter (reference cell type) to all methods of `QProjector`. In a follow-up PR (on which I am working), I will use this parameter to project quadrature rules onto faces of triangles and tetrahedrons.

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -45,6 +45,52 @@ namespace ReferenceCell
...
+          return Type::Line;
+        case 2:
+          return Type::Quad;
+        case 3:
+          return Type::Hex;
+        default:
+          Assert(false, ExcNotImplemented());
+          return Type::Invalid;
+      }
+  }
```
Comment (peterrum) [inline include/deal.II/grid/reference_cell.h:92]: To be able to program dimension independently I have introduced the following two functions.

File: include/deal.II/base/qprojector.h [hunk 3]
```diff
@@ -87,19 +89,40 @@ class QProjector
    * is used on face <tt>face_no</tt>. For further details, see the general
    * doc for this class.
    */
-  static void
+  DEAL_II_DEPRECATED static void
```
Comment (kronbichler) [inline include/deal.II/base/qprojector.h:92]: Can you add a statement in the function documentation suggesting that the function of the same name with a `ReferenceCell::Type` argument should be used?

File: include/deal.II/base/qprojector.h [hunk 4]
```diff
@@ -87,19 +89,40 @@ class QProjector
...
+                  std::vector<Point<dim>> & q_points);
+
   /**
    * Compute the cell quadrature formula corresponding to using
    * <tt>quadrature</tt> on face <tt>face_no</tt>. For further details, see
    * the general doc for this class.
    */
-  static Quadrature<dim>
+  DEAL_II_DEPRECATED static Quadrature<dim>
   project_to_face(const SubQuadrature &quadrature, const unsigned int face_no);
```
Comment (kronbichler) [inline include/deal.II/base/qprojector.h:114]: Same here and in all occurrences below.

File: include/deal.II/base/qprojector.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/base/qprojector.h:96]: Done!

**Pull request merged**